File: 백준/Gold/1113. 수영장 만들기/수영장 만들기.py
from collections import deque

n, m = map(int, input().split())
pool = {}
heights = []

for h in range(2, 10):
    pool[h] = { "walls": set(), "voids": set() }

for y in range(n):
    floor = input()
    for x in range(len(floor)):
        block = int(floor[x])
        position = (x, y)

        for h in range(2, 10):
            if block >= h:
                pool[h]["walls"].add(position)

            else:
                pool[h]["voids"].add(position)

moves = ( (0, 1), (0, -1), (1, 0), (-1, 0) )
way_count = {}

count = 0

for h in range(2, 10):

    walls = pool[h]["walls"]
    voids = pool[h]["voids"]

    while voids:
        is_border = False
        root = voids.pop()
        
        group = []
        group_count = 0
        
        queue = deque([root])
        while queue:
            block = queue.popleft()
            group.append(block)
            group_count += 1

            for move in moves:
                next = (block[0] + move[0], block[1] + move[1])

                if is_border is False:
                    if (next[0] < 0) or (next[0] > m - 1):
                        is_border = True

                    if (next[1] < 0) or (next[1] > n - 1):
                        is_border = True

                if next in voids:
                    voids.remove(next)
                    queue.append(next)

        if is_border is False:
            count += group_count

print(count)

File: 백준/Gold/1113. 수영장 만들기/수영장 만들기.cc
#include <bits/stdc++.h>

using namespace std;

int main(){

    ios::sync_with_stdio(false);

    cin.tie(0);

    

    int n, m;

    cin >> n >> m;

    

    vector<string> grid(n);

    for(int y = 0; y < n; ++y){

        cin >> grid[y];

    }

    

    vector<pair<int, int>> moves = { {0, 1}, {0, -1}, {1, 0}, {-1, 0} };

    

    long long count = 0;

    

    for(int h = 2; h <= 9; ++h){

        vector<vector<bool>> visited(n, vector<bool>(m, false));

        

        for(int y = 0; y < n; ++y){

            for(int x = 0; x < m; ++x){

                if((grid[y][x] - '0') < h && !visited[y][x]){

                    queue<pair<int, int>> q;

                    q.push({x, y});

                    visited[y][x] = true;

                    

                    bool is_border = false;

                    int group_count = 0;

                    

                    while(!q.empty()){

                        pair<int, int> current = q.front();

                        q.pop();

                        

                        int cx = current.first;

                        int cy = current.second;

                        group_count++;

                        

                        if(cx == 0 || cx == m-1 || cy == 0 || cy == n-1){

                            is_border = true;

                        }

                        

                        for(auto &move : moves){

                            int nx = cx + move.first;

                            int ny = cy + move.second;

                            

                            if(nx < 0 || nx >= m || ny < 0 || ny >= n){

                                is_border = true;

                                continue;

                            }

                            

                            if((grid[ny][nx] - '0') < h && !visited[ny][nx]){

                                visited[ny][nx] = true;

                                q.push({nx, ny});

                            }

                        }

                    }

                    

                    if(!is_border){

                        count += group_count;

                    }

                }

            }

        }

    }

    

    cout << count;

    

    return 0;

}